cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java

```diff
@@ -447,6 +487,20 @@ private int getInstanceIdFromColumn(byte[] column) {
   }
 
   @Nullable
+  private Map<Long, QueueBarrier> decodeBarrierInfo(Row row) {
+    Map<Long, QueueBarrier> barrierInfo = Maps.newHashMap();
+    Map<byte[], byte[]> columns = row.getColumns();
+    for (byte[] columnKey : columns.keySet()) {
+      long groupId = Bytes.toLong(columnKey);
+      QueueBarrier barrier = decodeBarrierInfo(row, groupId);
```

Comment:
That's a bit wasteful to encode groupId again and use it to get from row again. You can add a new method:

```java
QueueBarrier decodeBarrierInfo(byte[] rowKey, byte[] columnValue);
```

and have both existing `decodeBarrierInfo` call that.